python/TestHarness/testers/SignalTester.py

```diff
@@ -39,15 +39,16 @@ def send_signal(self,pid):
 
             #first, make a true duplicate of the stdout file so we don't mess with the seek on the actual file
             out_dupe = copy.copy(self.outfile)
-            #go to the beginning of the file and see if its actually started running the binary
-            out_dupe.seek(0)
-            output = out_dupe.read()
+            if not out_dupe.closed:
```

Comment:
I am very concerned that this works. I would expect the inability to read the file should cause this test to fail. But it does not. So I don't understand what is going on here.

Reply:
@pbehne

Reply:
Full disclosure: I worked on patching a bug with signal checkpoints, but didn't touch this send_signal function and don't think I fully understand it. That being said, shouldn't the time.sleep, os.kill, and break lines below be indented to be under the `if not out_dupe.closed` check? My assumption is that if out_dupe is already closed, then the moose binary has finished running. In that case, os.kill to a nonexistent PID should raise a ProcessLookupError. If this isn't happening, and the tests are always passing, my guess is that this code is always executed when the binary is running and out_dupe is not closed. I will poke around at this a bit more.

Reply:
[pid_example.txt](https://github.com/idaholab/moose/files/14030841/pid_example.txt)

I have simplified the logic into the attached script. The 'test' is running `echo test ; sleep 1`. When the script is run with line 37 commented and line 38 uncommented (i.e., `time.sleep(5)`), os.kill is still able to send a signal to the pid even though the process should have already been terminated (the timer gives a wait time > 5 s). However, when the script is run with line 37 uncommented and line 38 commented (i.e., `os.waitpid`), os.kill is not able to send the signal because the process has already terminated. HOWEVER, the time gives a wait time on the order of 1 s, so it seems reasonable that using `time.sleep(5)` without `waitpid` would give the subprocess PLENTY of time to terminate. I am not sure how to explain this behavior. Are you? @permcody?

Reply:
I myself don't know how to explain anything. All I know is that ever since I went to python 3.11 on my machine this test has failed every time I run `run_tests` and I'll be really happy when it doesn't fail anymore

Reply:
This may have to do with reaping an exited process. When a process terminates it doesn’t actually end. It hangs around until the parent process closes it out. Until that occurs. The “dead” process is left in “zombie” state. We can take a look at this tomorrow and get to the bottom of it.

Reply:
Well the test just failed on Intel Mac (https://civet.inl.gov/job/2016334/). It usually passes.

```pre
misc/signal_handler.test_signal_parallel: Working Directory: /opt/civet1/build/moose/test/tests/misc/signal_handler
misc/signal_handler.test_signal_parallel: Running command: mpiexec -n 3 /opt/civet1/build/moose/test/moose_test-opt -i simple_transient_diffusion_scaled.i Executioner/num_steps=70 --error --error-override --timing Outputs/perf_graph=true --no-gdb-backtrace
misc/signal_handler.test_signal_parallel: 
misc/signal_handler.test_signal_parallel: 
misc/signal_handler.test_signal_parallel: Exit Code: 30
misc/signal_handler.test_signal_parallel: ################################################################################
misc/signal_handler.test_signal_parallel: Tester failed, reason: CRASH
misc/signal_handler.test_signal_parallel: 
[1.090s]     FAIL misc/signal_handler.test_signal_parallel FAILED (CRASH) [min_cpus=3]
```